File: sorting/shell-sort/shell_sort.py
# -*- encoding: utf-8 -*-


def shell_sort(st_bf):
    length = len(st_bf)

    step = length / 2
    while step >= 1:
        print 'steps = {}'.format(step)
        for i in xrange(step, length):
            temp = st_bf[i]
            j = i - step
            while j >= 0 and st_bf[j] > temp:
                st_bf[j+step] = st_bf[j]
                j -= step
            st_bf[j+step] = temp
        print st_bf
        step = step / 2


st_bf = [6, 5, 3, 1, 8, 7, 2, 4, 2]
shell_sort(st_bf)

File: sorting/shell/shell-sort.cpp
#include <iostream>
using namespace std;


void print(int st_bf[], size_t size) {
    for (size_t i = 0; i < size; i++)
    {
        cout << st_bf[i] << " ";
    }
    cout << endl;
}


void shellsort(int *st_bf, size_t size) {
    for (size_t k = size/2; k >= 1; k = k/2) {
        cout << "k = " << k << endl;
        for (int i = k; i < size; ++i) {
            int temp = st_bf[i];
            int j = i - k;
            while (j >= 0 and st_bf[j] > temp) {
                st_bf[j+k] = st_bf[j];
                j -= k;
            }
            st_bf[j+k] = temp;
        }
        print(st_bf, size);
    }
}


int main(){
    int st_bf[] = {6, 5, 3, 1, 8, 7, 2, 4, 2};
    size_t size = sizeof(st_bf) / sizeof(st_bf[0]);
    shellsort(st_bf, size);
}
